### DISCUSSION TITLE
more precisely find dash button mac address?

### DISCUSSION THREAD

**Discussion by rpfile** [shanemcw/homebridge-amazondash-mac#22] [Ideas]
not sure if this output is specific to the macosx version of tcpdump or not, but i see that the mac address of the dash button is always preceded by SA: which i guess stand for 'source address'.

the regexp in the plugin finds the first mac address which is fine for the packets seen when a dash button probes for a known network name (but then fails to join.) however, if the dash button is configured with a correct wifi password, then sometimes it does not probe but instead just tries to join the network. the end result is that a different packet is emitted where the first mac address is a BSSID (and is preceded by BSSID:). in this case the plugin code misses the dash button's mac address (which is again preceded by SA:), and so misses the button press.

in my application i have another machine that looks for button presses by looking for the ARP that happens after the button joins the network, so i need my dash button to have a proper wifi config. of course my router drops all packets from its mac address and also blackholes the amazon dash button server IP, etc, so it's safe. i realize this is a configuration that most people might not want to deal with.

i've hacked the code to look for SA:xx:xx:xx:xx:xx:xx but in a somewhat inelegant way. if i could do it properly i could submit a PR, but i thought i would at least mention this problem for anyone else that comes across it.

**Comment by shanemcw** [comment]
Thanks — this makes sense. The current parser assumes the first MAC address on the line is the button, which is not always true for association traffic. Preferring the SA: address when present, with the existing MAC scan as a fallback for other capture formats, looks like the safer approach. New Dash provisioning is no longer a generally supported workflow, but this can still affect already-configured buttons.

I’m actively refactoring parts of the plugin now, so keep an eye out for an upcoming alpha release that addresses this.

**Reply by shanemcw** [reply]
https://npm.io/package/homebridge-amazondash-mac/v/4.0.0-beta.1

Take it out for a walk! LMK. Thanks for feedback!

**Reply by rpfile** [reply]
its been a while since i set this all up so i'll have to go reload the context into my brain (and avoid deleting the hacks i made, just in case)

**Reply by shanemcw** [reply]
@rpfile send me your local code with your hacks and I'll take a look! 

Here's the relevant section to your context, with the adjustment:
`let source = (self.dumpname === 'tcpdump') ? upperLine.match(/\bSA:((?:[\dA-F]{2}:){5}[\dA-F]{2})\b/) : null;`
